whylog/constraints/__init__.py

```diff
@@ -2,42 +2,180 @@
 
 import six
 
+from whylog.constraints.const import ConstraintType
+from whylog.constraints.exceptions import ConstructorGroupsError, ConstructorParamsError
+from whylog.teacher.user_intent import UserConstraintIntent
+
 
 @six.add_metaclass(ABCMeta)
 class AbstractConstraint(object):
+    """
+    :param MIN_GROUPS_COUNT: minimal groups count needed to create constraint
+    :param MAX_GROUPS_COUNT: maximal groups count needed to create constraint
+    """
+
+    TYPE = 'undefined'
+
+    MIN_GROUPS_COUNT = None
+    MAX_GROUPS_COUNT = None
+
+    PARAMS = []
+
     @abstractmethod
-    def save(self, rulebase_rule):
-        pass
+    def __init__(self, param_dict, groups):
+        """
+        For Teacher and Front use while creating user rule.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: all groups that are linked by constraint,
+                       represented by list of tuples (line_id, group_no),
+                       where line_id and group_no is inner numeration between Front and Teacher
+        """
+        self.params = param_dict
+        self.groups = groups
+        self._check_params()
+
+    def _check_params(self):
```

Comment:
This function doesn't check the params. It check only group count. Rename it or add checking constraint paramdict.